flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImplTest.java

```diff
@@ -0,0 +1,357 @@
+			new UnregisteredMetricsGroup(),
+			new UnregisteredMetricsGroup()
+		);
+	}
+
+	private static ResultPartitionDeploymentDescriptor createResultPartitionDeploymentDescriptor(ResultPartitionID resultPartitionId, boolean externallyManaged) {
+		return new ResultPartitionDeploymentDescriptor(
+			new PartitionDescriptor(
+				new IntermediateDataSetID(),
+				resultPartitionId.getPartitionId(),
+				externallyManaged
+					? ResultPartitionType.BLOCKING
+					: ResultPartitionType.PIPELINED,
+				1,
+				0),
+			(ShuffleDescriptor) () -> resultPartitionId,
+			1,
+			true);
+	}
+
+	private static class TestResultPartitionWriter implements ResultPartitionWriter {
+
+		private final ResultPartitionID resultPartitionID;
+
+		private TestResultPartitionWriter(ResultPartitionID resultPartitionID) {
+			this.resultPartitionID = resultPartitionID;
+		}
+
+		@Override
+		public void setup() throws IOException {
+
+		}
+
+		@Override
+		public ResultPartitionID getPartitionId() {
+			return resultPartitionID;
+		}
+
+		@Override
+		public int getNumberOfSubpartitions() {
+			return 0;
+		}
+
+		@Override
+		public int getNumTargetKeyGroups() {
+			return 0;
+		}
+
+		@Override
+		public BufferBuilder getBufferBuilder() throws IOException, InterruptedException {
+			return null;
+		}
+
+		@Override
+		public boolean addBufferConsumer(BufferConsumer bufferConsumer, int subpartitionIndex) throws IOException {
+			return false;
+		}
+
+		@Override
+		public void flushAll() {
+
+		}
+
+		@Override
+		public void flush(int subpartitionIndex) {
+
+		}
+
+		@Override
+		public void fail(@Nullable Throwable throwable) {
+
+		}
+
+		@Override
+		public void finish() throws IOException {
+
+		}
+
+		@Override
+		public void close() throws Exception {
+
+		}
+	}
+
+	private static class TestShuffleEnvironment implements ShuffleEnvironment<ResultPartitionWriter, InputGate> {
+
+		private final Collection<ResultPartitionWriter> resultPartitionWriters;
+		private final Collection<ResultPartitionID> releasedPartitions = new HashSet<>();
+
+		private TestShuffleEnvironment(Collection<ResultPartitionWriter> writersToReturn) {
+			this.resultPartitionWriters = writersToReturn;
+		}
+
+		Collection<ResultPartitionID> getReleasedPartitions() {
+			return releasedPartitions;
+		}
+
+		@Override
+		public int start() throws IOException {
+			return 1234;
+		}
+
+		@Override
+		public Collection<ResultPartitionWriter> createResultPartitionWriters(String taskName, ExecutionAttemptID executionAttemptID, Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors, MetricGroup outputGroup, MetricGroup buffersGroup) {
+			return resultPartitionWriters;
+		}
+
+		@Override
+		public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
+			releasedPartitions.addAll(partitionIds);
+		}
+
+		@Override
+		public Collection<ResultPartitionID> getPartitionsOccupyingLocalResources() {
+			return null;
```

Comment:
might better to return an empty collection if this interface method does not expect `null` return.